File: computer3/Fall Term/Solutions/Class11_Nov_20/lvl3_jayden_study_trees.py
import sys
input = sys.stdin.readline

n = int(input())
tree = [[] for i in range(n+1)]
for i in range(n-1):
    x, y = map(int, input().split())
    tree[x].append(y)
    tree[y].append(x)

far = 0
dis = [-1]*(n+1)
q = [1]
dis[1] = 0
while len(q) > 0:
    far = q.pop(0)
    for nxt in tree[far]:
        if dis[nxt] == -1:
            q.append(nxt)
            dis[nxt] = dis[far] + 1

dis = [-1]*(n+1)
q = [far]
dis[far] = 0
while len(q) > 0:
    far = q.pop(0)
    for nxt in tree[far]:
        if dis[nxt] == -1:
            q.append(nxt)
            dis[nxt] = dis[far] + 1

print(dis[far])

File: computer3/Fall Term/Solutions/Class11_Nov_20/lvl3_jayden_study_trees.java
import java.util.*;
import java.io.*;
public class Main {
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
	static StringTokenizer st;
	static int n, len, far; static List<Integer> adj[];
	public static void main(String[] args) throws IOException{	
		n = readInt(); adj = new ArrayList[n+1];
		for(int i=1; i<=n; i++) adj[i] = new ArrayList<>();
		for(int i=1; i<n; i++) {
			int u = readInt(), v = readInt();
			adj[u].add(v); adj[v].add(u);
		}
		dfs(1, -1, 0);  len=0; dfs(far, -1, 0);
		System.out.println(len);
	}
	static void dfs(int cur, int pre, int dis) {
		if(dis > len) { far = cur; len = dis; }
		for(int nxt : adj[cur]){
			if(nxt != pre) dfs(nxt, cur, dis+1);
		}
	}
	static String next () throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	static long readLong () throws IOException {
		return Long.parseLong(next());
	}
	static int readInt () throws IOException {
		return Integer.parseInt(next());
	}
	static double readDouble () throws IOException {
		return Double.parseDouble(next());
	}
	static char readCharacter () throws IOException {
		return next().charAt(0);
	}
	static String readLine () throws IOException {
		return br.readLine().trim();
	}
}